src/AggregateFunctions/AggregateFunctionGroupArray.h

```diff
@@ -266,19 +266,20 @@ class GroupArrayNumericImpl final
     void serialize(ConstAggregateDataPtr __restrict place, WriteBuffer & buf, std::optional<size_t> /* version */) const override
     {
         const auto & value = this->data(place).value;
-        size_t size = value.size();
+        const size_t size = value.size();
         writeVarUInt(size, buf);
-        buf.write(reinterpret_cast<const char *>(value.data()), size * sizeof(value[0]));
+        for (const auto & element : value)
```

Comment:
This is replacing a single write call with `size` write calls. While this might be necessary in big-endian machines, I expect this will make things slower for little-endian ones, which are the most common ones.

Same for reading. If you are in a little-endian arch you should be able to do the fast path of simply copying whole blocks with one operation. CH is fast because it does things in blocks :smile:

Reply:
Serialization and deserialization (both affected here) gets invoked all the time when you work with materialized views, projections or distributed queries (any query that needs to either store, retrieve or send aggregation states).

Reply:
The reason I went with this approach is I encountered something similar in another PR (https://github.com/ClickHouse/ClickHouse/pull/50276/files#r1216984994) where @rschu1ze suggested that instead of reading the whole thing into memory and swapping items in case of big-endian, I could go one-by-one. I actually agreed with that sentiment, since in terms of maintainability going one-by-one is somewhat superior.

With that being said, I could also go with the approach of reading everything then swapping them in case of big-endian if the impact on performance was significant.

Reply:
It depends on the kind of queries run how often serialization/deserialization is invoked. In this particular case, I would argue that aggregate function `groupArray` is rather exotic, it is not official SQL (though many DBMS seem to have it in one form or the other). Interestingly, there are a few performance tests for `groupArray` in ClickHouse, so there is a chance that performance regressions are caught.

How about this: we'll see if the performance tests complain (my bet is no), and if they do, we introduce sth. like this:

```cpp
if constexpr (std::endian::native == std::endian::little)
    buf.write(reinterpret_cast<const char *>(value.data()), size * sizeof(value[0]));
else
    for (const auto & element : value)
        writeBinaryLittleEndian(element, buf);
```

(ideally along with some comment)

Reply:
> groupArray is rather exotic

I see groupArray used quite often to replace window functions in materialized views.

> here are a few performance tests for groupArray in ClickHouse

Not sure if they are testing serialization and deserialization though.

But this is beside the point IMO. It is more about setting up a good foundation to do further changes in the future so that serialization is truly endianness independent and you can take data written by a CH and use it in another without problems. If this is done just so a few tests pass then it's the wrong reason as a) this is a breaking change (which BTW should be tagged as such for big-endianness platforms as they won't be able to read states written in the previous release) and b) I'd expect most of the serialization irks to not be covered by any tests. 

Finally, about loops vs single batches: [here is an example](https://github.com/ClickHouse/ClickHouse/pull/41089/) where changing a loop to a single read() call made uniqMerge 2x faster. These things do matter in some situations.

Reply:
Your point around breaking change is a valid one, but the official stance is: https://github.com/ClickHouse/ClickHouse/pull/48986#issuecomment-1517038580.

The main reasons for going gradually about introducing support for interoperability is two-fold:
* multiple people are working on this so this way the effort is divided into smaller pieces,
* having to do it all at once would result in a terribly long change that'd be much harder to review.

All of this is also enabled by the fact that big-endian isn't officially supported, so backward-compatibility isn't a concern until everything comes together.